### PR TITLE
[TimerTrigger] fixing daylight savings bug

### LINKED ISSUE #880: TimerTrigger issue with Daylight Saving Time change
Daylight saving time change caused the timer trigger not to trigger for 1hour for below repro app.

#### Repro app details -

Function App : prod-mcz-balance-app
Timer trigger: SendImbalanceToFiftyoneAsync
OS: Windows
App setting 'WEBSITE_TIME_ZONE' is set to: 'Romance Standard Time'

When Daylight saving time was set 1 hour back on 29th of October the TimerTrigger did not trigger for 1 hour between 12.00 AM - 01.00 AM UTC. 

#### Related information 

https://github.com/Azure/azure-webjobs-sdk-extensions/pull/492#issuecomment-479862622

### PR DESCRIPTION
**Pull request by brettsam**
Fixes #880

This change refactors our handling of Daylight Savings Time (DST). Previously this was all done in `GetNextTimerInterval()`, but after researching and testing, I realized that there wasn't enough information available to correctly calculate Cron schedules during the transition into and out of DST. Instead, I've moved the DST handling into the `CronSchedule` (and `ConstantSchedule`, although this one was easier) itself. This has the added benefit that when we print out the next several occurrences at the beginning of the Timer startup, they'll now be correct. Previously, we'd generate the "next" value and then re-calculate it based on DST to start our internal timer.

I've tried to add comments with explanations of terms to the code itself but have called a few things out in comments as well.

Side note that this only touches `Cron` and `Constant` schedules, which are the only ones supported by Functions. I briefly tested the behavior of `Daily` and `Weekly` and they seem... fine? It's hard to know exactly how they should behave, but they're so little used that it wasn't worth the effort to re-think them.

Some other notes: 
- `DateTime` can *sometimes* store enough details internally to allow it to roundtrip to/from `DateTimeOffset`. I've changed our usage to start from `DateTimeOffset.Now` wherever possible to ensure we have all the data for offsets stored internally. 
- However, NCronTab internally does *not* return `DateTime`s with offset details. This means that in ambiguous times (times during "fall back" from DST where there are duplicate times in a time zone), we will always think that this is the Standard (later) offset because of how `GetUtcOffset()` works. I've added calculations for the `CronSchedule` to account for this.
    > Side note. `GetUtcOffset()` only returns the Standard time in ambiguous cases when it doesn't have enough internal information to choose the correct offset. Sometimes it does! Like if you use `dateTimeOffset.LocalDateTime`, or even `DateTime.Now` it [stores some details internally](https://github.com/dotnet/runtime/blob/f722e5c5b72000a06ce267a9bee222a8c3be4696/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L133-L137) that let it choose the correct offset in ambiguous cases.
- I had previously fixed an issue when "springing forward" to DST where we'd calculate the wrong interval. In my testing, I found some other incidents where we'd schedule a trigger for a nonexistent time (i.e 02:30 when springing forward from 01:00 -> 03:00). I've taken this into account now as well.
- For `CronSchedule`, the behavior is different when you have an "interval" or a "point-in-time" schedule as the behavior around the "ambiguous" times going from DST -> Standard are different. I followed similar behavior to this: https://github.com/HangfireIO/Cronos?tab=readme-ov-file#transition-from-summer-time-in-autumn.

A succinct example of what this fixes can be shown here in this log for when daylight savings exits:

before change:
```
The next 5 occurrences of the 'TimerTrigger' schedule (Cron: '0 0,30 * * * *') will be:
      11/03/2024 00:30:00-07:00 (11/03/2024 07:30:00Z)
      11/03/2024 01:00:00-08:00 (11/03/2024 09:00:00Z)
      11/03/2024 01:30:00-08:00 (11/03/2024 09:30:00Z)
      11/03/2024 02:00:00-08:00 (11/03/2024 10:00:00Z)
      11/03/2024 02:30:00-08:00 (11/03/2024 10:30:00Z)
```

after change:
```
The next 5 occurrences of the 'TimerTrigger' schedule (Cron: '0 0,30 * * * *') will be:
      11/03/2024 00:30:00-07:00 (11/03/2024 07:30:00Z)
      11/03/2024 01:00:00-07:00 (11/03/2024 08:00:00Z)
      11/03/2024 01:30:00-07:00 (11/03/2024 08:30:00Z)
      11/03/2024 01:00:00-08:00 (11/03/2024 09:00:00Z)
      11/03/2024 01:30:00-08:00 (11/03/2024 09:30:00Z)
```

### CODE REVIEW

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 1]
```diff
@@ -32,13 +51,103 @@ internal CrontabSchedule InnerSchedule
...
+            // Scenario 1 -- When Daylight Savings begins and the clock springs forward, we need to ensure that we are 
+            //   not scheduling a job to run at an invalid time. If the next occurrence is invalid, we need to skip
+            //   it and look for the next possible valid time.
+            bool isInvalidTime = TimeZone.IsInvalidTime(DateTime.SpecifyKind(nextDateTime, DateTimeKind.Unspecified));
+
+            while (isInvalidTime)
+            {
+                nextDateTime = _cronSchedule.GetNextOccurrence(nextDateTime);
+                isInvalidTime = TimeZone.IsInvalidTime(DateTime.SpecifyKind(nextDateTime, DateTimeKind.Unspecified));
+            }
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Does it matter when the next occurrence happens after the invalid occurrences are discarded?

Or is this a case of only matters if interval and doesn't matter if "point-in-time"?

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 2]
```diff
@@ -32,13 +51,103 @@ internal CrontabSchedule InnerSchedule
...
+            // Begin evaluating scenarios when Daylight Savings ends and time falls back. This leads to "ambiguous" times
+            // as the clock repeats an hour. For example, times from 1:00 - 1:59 AM will occur twice in Pacific Standard Time
+            // and are therefore considered "ambiguous" for this time zone.            
+            bool isNowAmbiguous = TimeZone.IsAmbiguousTime(nowDateTime);
+            bool isNextAmbiguous = TimeZone.IsAmbiguousTime(nextDateTime);
+
+            // Because of how NCronTab constructs it's "next" DateTime, if the next time is ambiguous, it will *always*
+            // return the Standard Time offset when calling GetUtcOffset().
+            var nextOffset = TimeZone.GetUtcOffset(nextDateTime);
+            var next = new DateTimeOffset(nextDateTime, nextOffset);
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: What is the specific scenario if we end here? What is `next` at this point? (i.e. all the if-else statements below don't evaluate, and we go straight to `return next;`) Is this still scenario 1 described above?

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 3]
```diff
@@ -32,13 +51,103 @@ internal CrontabSchedule InnerSchedule
...
+            {
+                // Scenario 4 -- Interval schedule where either point is ambiguous and we're crossing the offset boundary. For example, 
+                //   an "every 30 minute" interval in Pacific time:
+                //   - 00:30-7 -> 01:00-8 (because all ambiguous "next" use Standard time) -> adjust back 01:00-7
+                //   - 01:30-7 -> 02:00-8 -> adjust back to 01:00-8
+                var offsetDiff = nextOffset - now.Offset;
+                next = TimeZoneInfo.ConvertTime(next.Add(offsetDiff), TimeZone);
+            }
+
+            return next;
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: I think this all mostly makes sense to me; I'm just struggling to wrap my head around the difference in the "spring forward" vs "fall back" calculations since they're all happening in this one function i.e. we're checking for DST in both directions every time which is harder to follow for me

**Comment by liliankasem** [comment]
I just want to call out that the code comments provided here are fantastic, I was dreading to look at this PR but the descriptions were concise yet detailed enough. It was really easy to follow along and understand the changes being made, and why!

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 2 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: I guess technically this is really `Scenario 0` -- where we're not in any kind of invalid or ambiguous time frame and the time that we get back from NCronTab simply needs to be created using the non-confusing TimeZone offset. I'll add comments for that...

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 3 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Yes, I agree. Let me break them out to make this more clear. I realized that as I was reading through it again also.

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 4]
```diff
@@ -82,28 +86,72 @@ private DateTimeOffset GetNextOccurrence(DateTimeOffset now)
...
+            if (TryAdjustInvalidTime(next, out DateTimeOffset? adjusted))
+            {
+                // "Spring Forward": "next" was invalid and was adjusted forward to a valid time.
+                next = adjusted.Value;
+            }
+            else if (TryAdjustAmbiguousTime(now, next, out adjusted))
+            {
+                // "Fall Back": Either "now" or "next" was ambiguous and the time was adjusted.
+                next = adjusted.Value;
+            }
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Soooo much cleaner!

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 5]
```diff
@@ -82,28 +86,72 @@ private DateTimeOffset GetNextOccurrence(DateTimeOffset now)
...
                 nextDateTime = _cronSchedule.GetNextOccurrence(nextDateTime);
-                isInvalidTime = timeZone.IsInvalidTime(DateTime.SpecifyKind(nextDateTime, DateTimeKind.Unspecified));
+                isInvalidTime = TimeZoneInfo.Local.IsInvalidTime(nextDateTime);
+            }
+
+            adjusted = new DateTimeOffset(nextDateTime);
+            return true;
+        }
+
+        internal bool TryAdjustAmbiguousTime(DateTimeOffset now, DateTimeOffset next, out DateTimeOffset? adjusted)
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs:125]: Can you add a summary comment here too?

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 6]
```diff
@@ -82,28 +86,72 @@ private DateTimeOffset GetNextOccurrence(DateTimeOffset now)
...
+        {
+            adjusted = null;
+
+            TimeZoneInfo timeZone = TimeZoneInfo.Local;
+            if (!timeZone.SupportsDaylightSavingTime)
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Can there be an invalid time if there is no DST? Do we need to add this check to ` TryAdjustInvalidTime` as well?Maybe it makes more sense to have this check be part of `GetNextOccurrence` before we do the "TryAdjust" bits

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 7]
```diff
@@ -116,38 +164,40 @@ private DateTimeOffset GetNextOccurrence(DateTimeOffset now)
...
                 //   - 00:30-7 -> 01:00-8 (because all ambiguous "next" use Standard time) -> adjust back 01:00-7
                 //   - 01:30-7 -> 02:00-8 -> adjust back to 01:00-8
-                var offsetDiff = nextOffset - now.Offset;
-                next = TimeZoneInfo.ConvertTime(next.Add(offsetDiff), timeZone);
+                var offsetDiff = next.Offset - now.Offset;
+                adjusted = TimeZoneInfo.ConvertTime(next.Add(offsetDiff), timeZone);
             }
 
-            return next;
+            return adjusted != null;
```
Comment (liliankasem) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: nit: I'd like to see us using the `is` operator more throughout our repos for improved code readability and consistency

```suggestion
            return adjusted is not null;
```

**Review by liliankasem — APPROVED** [review]
lgtm, minor comments

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 8]
```diff
@@ -32,13 +61,149 @@ internal CrontabSchedule InnerSchedule
...
+        internal bool TryAdjustInvalidTime(DateTimeOffset next, out DateTimeOffset? adjusted)
+        {
+            adjusted = null;
+
+            if (TimeZoneInfo.Local.IsInvalidTime(next.DateTime))
+            {
+                // This forces the offset to be recalculated for the local time zone. It will also signal that we don't need to do the
+                // ambiguous time calculations as this cannot happen during ambiguous times.
+                adjusted = next.ToLocalTime();
+                _logger.LogDebug(new EventId(900, "DstInvalidTime"), $"The calculated next occurrence for the schedule '{_cronSchedule}' of '{next}' was determined to be an invalid time in the time zone '{TimeZoneInfo.Local.DisplayName}' due to Daylight Saving Time. The adjusted next occurrence is '{adjusted}'.");
```
Comment (cjaliaga) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Would make sense to have these with `LoggerMessage.Define` (on a partial class or maybe using extension methods) and using structured logging? Not a hot path (well maybe for one hour a year) but I like having them listed with their EventId/EventName one after the other.

Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Sure, I moved them to their own Logger class like others.

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 6 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: I actually just refactored that as well -- this was just an optimization, but I moved it up to GetNextOccurrence now (as you suggested) so it's clearer that if there's no DaylightSavingTime, nothing else needs to happen.

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 9]
```diff
@@ -32,13 +61,149 @@ internal CrontabSchedule InnerSchedule
...
+        /// <summary>
+        /// If the provided DateTime is not valid, moves forward to the next occurrence in the schedule
+        /// until a valid DateTime is found. For example, 02:30 does not exist when "springing forward"
+        /// to Daylight Saving Time in the Pacific Standard Time Zone. Uses TimeZoneInfo.Local to make
+        /// the determination.
+        /// </summary>
+        /// <param name="next">The DateTimeOffset to evaluate.</param>
+        /// <param name="adjusted">The adjusted DateTimeOffset, if "next" is invalid.</param>
+        /// <returns></returns>
+        internal bool TryAdjustInvalidTime(DateTimeOffset next, out DateTimeOffset? adjusted)
```
Comment (brettsam) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: I realized that this was completely unnecessary. Just calling `next.ToLocalTime()` takes care of any invalid time that may have been created. So I've removed the entire method in the next iteration.

File: src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs [hunk 10]
```diff
@@ -94,8 +94,9 @@ public async Task StartAsync(CancellationToken cancellationToken)
             bool isPastDue = false;
 
             // we use DateTime.Now rather than DateTime.UtcNow to allow the local machine to set the time zone. In Azure this will be
-            // UTC by default, but can be configured to use any time zone if it makes scheduling easier.
-            DateTime now = DateTime.Now;
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs:98]: nit: fix DateTime.Now comments above, updating them with DateTimeOffset

File: src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs [hunk 11]
```diff
@@ -266,7 +267,7 @@ internal async Task HandleTimerEvent()
         /// <param name="invocationTime">The time of the invocation, likely DateTime.Now.</param>
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs]: DateTimeOffset.Now

File: src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs [hunk 12]
```diff
@@ -145,7 +146,7 @@ public async Task StartAsync(CancellationToken cancellationToken)
             else
             {
                 // start the regular schedule
-                StartTimer(DateTime.Now);
+                StartTimer(DateTimeOffset.Now);
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs:149]: I wonder why we've never used the "now" local captured above here?

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/TimerSchedule.cs [hunk 13]
```diff
@@ -19,6 +18,7 @@ public abstract class TimerSchedule
         /// <summary>
         /// Gets a value indicating whether intervals between invocations should account for DST.        
         /// </summary>
+        [Obsolete("This property is obsolete and will be removed in a future version. All TimerSchedule implementations should now handle their own DST transitions.")]
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/TimerSchedule.cs:21]: Does this show up in the .NET Isolated mapping interfaces?

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 14]
```diff
@@ -14,14 +16,41 @@ namespace Microsoft.Azure.WebJobs.Extensions.Timers
     public class CronSchedule : TimerSchedule
     {
         private readonly CrontabSchedule _cronSchedule;
+        private readonly ILogger _logger;
+
+        internal CronSchedule(CrontabSchedule schedule, ILogger logger)
+        {
+            _cronSchedule = schedule;
+            var cron = schedule.ToString();
+            var parts = cron.Split(' ');
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: I think you may want to split like the HasSeconds ensuring we remove empty entries. E.g. maybe add a test case like "     
          * * * * * *" to your tests and see what happens?

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 15]
```diff
@@ -32,13 +61,133 @@ internal CrontabSchedule InnerSchedule
...
         public override DateTime GetNextOccurrence(DateTime now)
         {
-            return _cronSchedule.GetNextOccurrence(now);
+            // Note: TimeZoneInfo.Local is mocked in tests
+            return GetNextOccurrence(new DateTimeOffset(now)).LocalDateTime;
+        }
+
+        private DateTimeOffset GetNextOccurrence(DateTimeOffset now)
+        {
+            // The cron library only supports DateTime and does not take TimeZones into account. We need to
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs:79]: Looks like there's an ncrontab version 3.3.3 with bug fixes that we might consider taking. I only checked to see if there was some updated version with TimeZone support :)

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 16]
```diff
@@ -32,13 +61,133 @@ internal CrontabSchedule InnerSchedule
...
+                // "Fall Back": Either "now" or "next" was ambiguous and the time was adjusted.
+                next = adjusted.Value;
+            }
+
+            return next;
+        }
+
+        /// <summary>
+        /// Uses the 'now' and 'next' values to determine if either is ambiguous. If they are, adjusts the times
+        /// so that they are unambiguous, depending on the time of schedule it is (interval or point-in-time).
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: "type" of schedule it is

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 17]
```diff
@@ -32,13 +61,133 @@ internal CrontabSchedule InnerSchedule
...
+            //    - Run on November 3rd at 01:30-7
+            //    - Run on November 4th at 01:30-7 (and not run at 1:30-8)
+            //    - Run on November 5th at 01:30-8
+            if (!IsInterval && isNextAmbiguous && next.Offset != now.Offset)
+            {
+                // "Fall Back" Scenario 1 -- Point-in-time schedule where the next time is ambiguous and the offsets have changed.
+                //   This means that "now" is in DST and "next" is in Standard Time. Since we never want to run an ambiguous
+                //   point-in-time schedule on Standard time, move the offset back to the Daylight offset.
+                var offsetDiff = next.Offset - now.Offset;
+                adjusted = TimeZoneInfo.ConvertTime(next.Add(offsetDiff), timeZone);
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs:168]: I'm having a hard time following this offset adjustment logic. Can you tie it to your example above, or update the example above to explicitly give examples for these two different scenarios? Or maybe just include a "For example" here to make it clear.

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs [hunk 18]
```diff
@@ -32,13 +61,133 @@ internal CrontabSchedule InnerSchedule
...
+                Logger.DstAmbiguousTimeSecondOccurrence(_logger, now, _cronSchedule.ToString(), next, TimeZoneInfo.Local.DisplayName, adjusted.Value);
+            }
+            else if (IsInterval && (isNowAmbiguous || isNextAmbiguous) && next.Offset != now.Offset)
+            {
+                // "Fall Back" Scenario 3 -- Interval schedule where either point is ambiguous and we're crossing the offset boundary. For example, 
+                //   an "every 30 minute" interval in Pacific time:
+                //   - 00:30-7 -> 01:00-8 (because all ambiguous "next" use Standard time) -> adjust back 01:00-7
+                //   - 01:30-7 -> 02:00-8 -> adjust back to 01:00-8
+                var offsetDiff = next.Offset - now.Offset;
+                adjusted = TimeZoneInfo.ConvertTime(next.Add(offsetDiff), timeZone);
```
Comment (mathewc) [inline src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs:201]: To be sure I follow, you're saying:

next=01:00-8
now=00:30-7
offsetDiff=8-7=1
adjusted=01:00-8 + 1 = 01:00-7

That example makes sense to me. But I don't see how you get 01:00-8 from 01:30-7. Is TimeZoneInfo.ConvertTime doing this internally for you?

File: src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs (hunk 12 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Listener/TimerListener.cs:149]: Hmm, yeah, me too. Maybe because when we first captured it and now could make the interval incorrect? I'm hesitant to touch this part at this point...

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/TimerSchedule.cs (hunk 13 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/TimerSchedule.cs:21]: No -- looks like we never added the `Schedule` to that payload. I think we were trying to map the fewest required properties that we could?

https://github.com/Azure/azure-functions-dotnet-worker/blob/main/extensions/Worker.Extensions.Timer/src/TimerInfo.cs

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 14 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Good callout. I tested this just now and turns out we don't explicitly need to do that b/c we're looking at this after `CrontabSchedule` has already done all of its internal normalization. But I'm adding it anyway.

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 15 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs:79]: Yeah, I checked too :-)

Good point, I'll pull it in.

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 18 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs:201]: `DateTimeOffset` represents an exact point in time, UTC, and the `Offset` is giving you a Local view of that time, but it knows nothing about the Time Zone or if the time is valid, for example, because none of that information exists in the way that NCronTab creates the "next".

So, during ambiguous times, the current view we have is always of the Standard/later time. In this case, with a -8 offset.

By adding the offset difference to the time (which in this case `(-8) - (-7) == -1` is actually subtracting an hour), we're subtracting one hour from the UTC point-in-time and telling it to give us the Offset based on the local time zone of that UTC point-in-time.

Put simply, we're an hour late and we want to move it back. But just moving it back doesn't adjust the internal offset (again, b/c it doesn't know about time zones inherently). So, we use `TimeZoneInfo.ConvertTime()` to get the offset correct. This is really just for outputting logs, though -- because I believe that just subracting an hour would work for all our interval calculations. But it makes the output look odd and it would show a time that doesn't actually exist in that time zone. (`11/4/2018 00:30-8` in this case isn't a real time; it should be `11/4/2018 01:30-7`).

File: src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs (hunk 1 above)
Comment (brettsam) [reply src/WebJobs.Extensions/Extensions/Timers/Scheduling/CronSchedule.cs]: Removed this invalid time stuff... it's all handled differently now.

**Review by mathewc — APPROVED** [review]
